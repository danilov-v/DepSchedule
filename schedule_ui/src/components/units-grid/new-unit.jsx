import React from "react";
import classnames from "classnames";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

export function NewUnit({ onAddUnit }) {
  return (
    <div
      className={classnames(
        "unit",
        "d-flex",
        "justify-content-center",
        "align-items-center",
        "text-center",
        "unit-last-generation"
      )}
      onClick={() => onAddUnit()}
    >
      <FontAwesomeIcon icon="plus" />
    </div>
  );
}
