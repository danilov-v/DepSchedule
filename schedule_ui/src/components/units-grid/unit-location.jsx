import React from "react";

export function UnitLocation({ location, flag, onEditUnit }) {
  return (
    <div
      className="unit-location d-flex align-items-center justify-content-center"
      onClick={onEditUnit}
    >
      {flag && <img src={flag} alt="Unit flag preview" className="unit-flag" />}
      {location}
    </div>
  );
}
