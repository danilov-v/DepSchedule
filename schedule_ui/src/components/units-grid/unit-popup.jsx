import React, { useState, useRef } from "react";
import PropTypes from "prop-types";
import {
  Button,
  Form,
  FormGroup,
  Label,
  Input,
  Modal,
  ModalHeader,
  ModalBody,
  ModalFooter,
  DropdownMenu,
  DropdownItem,
  UncontrolledDropdown,
  DropdownToggle,
} from "reactstrap";
import { get, isBoolean } from "lodash";
import { createUnit, updateUnit } from "helpers/api";
import {
  SUCCESS_UNIT_NOTIFICATION_DATA,
  SUCCESS_UNIT_NOTIFICATION_DATA_EDIT,
  FAILED_UNIT_NOTIFICATION_DATA,
} from "constants/notifications";
import { FLAGS_MAP, NO_FLAG } from "constants/flags";
import { NotificationManager } from "helpers/notification-manager";

import "./unit-popup.scss";

const DEFAULT_FORM_DATA = {
  title: "",
  parentId: 0,
  flag: NO_FLAG.url,
  exists: false,
};

const validate = formData =>
  Boolean(formData.title && formData.title.length >= 2);

export function UnitPopup({ units, unit, isEdit, onUnitsUpdate, onClose }) {
  const [isValid, setValidState] = useState(null);
  const [formData, setData] = useState(unit);
  const titleInputEl = useRef(null);
  const toggleModal = () => {
    onClose();
  };
  const onInputChange = event => {
    const newFormData = { ...formData, title: get(event, "target.value") };

    setValidState(validate(newFormData));
    setData(newFormData);
  };
  const onSelectChange = event =>
    setData({ ...formData, parentId: +get(event, "target.value") });

  const onSubmit = async event => {
    event.preventDefault();

    if (validate(formData)) {
      try {
        if (isEdit) {
          await updateUnit(formData);
        } else {
          await createUnit(formData);
        }

        onUnitsUpdate();
        toggleModal();

        NotificationManager.fire(
          isEdit
            ? SUCCESS_UNIT_NOTIFICATION_DATA_EDIT
            : SUCCESS_UNIT_NOTIFICATION_DATA
        );
      } catch (e) {
        setValidState(false);
        NotificationManager.fire(FAILED_UNIT_NOTIFICATION_DATA);
      }
    } else {
      setValidState(false);
      NotificationManager.fire(FAILED_UNIT_NOTIFICATION_DATA);
    }

    return false;
  };

  const onEnter = () => {
    setTimeout(() => {
      if (titleInputEl.current) titleInputEl.current.focus();
    });
  };

  const onFlagChange = flag => {
    const { url } = flag;
    const newFormData = { ...formData, flag: url };

    setData(newFormData);
  };

  const onPlannedChange = event => {
    const { checked } = event.target;

    setData({ ...formData, exists: checked });
  };

  return (
    <Modal
      className="unit-popup"
      isOpen={true}
      toggle={toggleModal}
      onEnter={onEnter}
    >
      <Form className="p-3" onSubmit={onSubmit}>
        <ModalHeader toggle={toggleModal}>
          {isEdit ? "Редактирование" : "Создание"} Подразделения
        </ModalHeader>
        <ModalBody>
          <FormGroup>
            <Label for="unitTitle">Название Подразделения</Label>
            <Input
              name="unitTitle"
              id="unitTitle"
              placeholder="СУ"
              valid={isValid}
              invalid={isBoolean(isValid) && !isValid}
              onChange={onInputChange}
              value={formData.title}
              innerRef={titleInputEl}
            />
          </FormGroup>
          <FormGroup>
            <Label for="unitParent">Родительское Подразделение</Label>
            <Input
              type="select"
              name="unitParent"
              id="unitParent"
              onChange={onSelectChange}
              defaultValue={formData.parentId}
            >
              <option value={0}>Новое Подразделение</option>
              {units.map((unit, i) => (
                <option key={unit.title + i} value={unit.unitId}>
                  {unit.title}
                </option>
              ))}
            </Input>
          </FormGroup>
          <FormGroup>
            <UncontrolledDropdown>
              <DropdownToggle tag="label">
                Флаг Подразделения
                <div>
                  {formData.flag ? (
                    <img
                      src={formData.flag}
                      className="unit-popup-flag-item"
                      alt="flag preview"
                    />
                  ) : (
                    <span className="font-italic">{NO_FLAG.text}</span>
                  )}
                </div>
              </DropdownToggle>

              <DropdownMenu>
                <DropdownItem
                  value={NO_FLAG.id}
                  onClick={() => onFlagChange(NO_FLAG)}
                >
                  <span>{NO_FLAG.text}</span>
                </DropdownItem>
                {FLAGS_MAP.map(flag => (
                  <DropdownItem
                    key={flag.id}
                    value={flag.id}
                    onClick={() => onFlagChange(flag)}
                  >
                    <img
                      src={flag.url}
                      className="unit-popup-flag-item"
                      alt="flag drop-down-preview"
                    />
                  </DropdownItem>
                ))}
              </DropdownMenu>
            </UncontrolledDropdown>
          </FormGroup>
          <FormGroup check>
            <Label for="unitPlanned" check>
              <Input
                type="checkbox"
                name="unitPlanned"
                id="unitPlanned"
                onChange={onPlannedChange}
              />
              Запланировано
            </Label>
          </FormGroup>
        </ModalBody>
        <ModalFooter>
          <Button type="submit" color="success" className="mr-3">
            {isEdit ? "Обновить" : "Создать"}
          </Button>
          <Button color="primary" onClick={toggleModal}>
            Закрыть
          </Button>
        </ModalFooter>
      </Form>
    </Modal>
  );
}

UnitPopup.propTypes = {
  unit: PropTypes.shape({
    unitId: PropTypes.number,
    parentId: PropTypes.number,
    title: PropTypes.string,
  }),
  onClose: PropTypes.func,
};

UnitPopup.defaultProps = {
  unit: DEFAULT_FORM_DATA,
  onClose: () => {},
};
