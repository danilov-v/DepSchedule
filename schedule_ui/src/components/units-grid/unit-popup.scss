.unit-popup {
  &-flag-item {
    height: 40px;
  }

  .dropdown {
    label {
      cursor: pointer;
    }

    &-menu {
      overflow-y: scroll;
      max-height: 400px;
    }
  }
}
