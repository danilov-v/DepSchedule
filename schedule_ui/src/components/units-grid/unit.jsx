import React from "react";
import PropTypes from "prop-types";
import { Row, Col } from "reactstrap";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import classnames from "classnames";

export function Unit({
  title,
  unitId,
  lastGen,
  lastOfGroup,
  onAddUnit,
  onEditUnit,
  onRemoveUnit,
}) {
  const addUnit = () => onAddUnit(unitId);
  const editUnit = () => onEditUnit({ unitId, title });
  const removeUnit = () => onRemoveUnit(unitId);
  const iconClass =
    "unit-action d-flex align-items-center justify-content-center h-100";
  return (
    <div
      className={classnames(
        "unit",
        "d-flex",
        "justify-content-center",
        "align-items-center",
        "text-center",
        { "unit-last-generation": lastGen },
        { "unit-last-of-group": lastOfGroup }
      )}
      data-unit-id={unitId}
    >
      {title}
      <Row className="unit-actions" noGutters>
        <Col xs={4} className={iconClass} onClick={addUnit}>
          <FontAwesomeIcon icon="plus" />
        </Col>
        <Col xs={4} className={iconClass} onClick={editUnit}>
          <FontAwesomeIcon icon="edit" />
        </Col>
        <Col xs={4} className={iconClass} onClick={removeUnit}>
          <FontAwesomeIcon icon="trash-alt" />
        </Col>
      </Row>
    </div>
  );
}

Unit.propTypes = {
  title: PropTypes.string.isRequired,
  unitId: PropTypes.number.isRequired,
  lastGen: PropTypes.bool,
};
