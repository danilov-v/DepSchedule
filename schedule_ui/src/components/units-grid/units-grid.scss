.units-grid {
  &-title {
    height: 175px;
    background: #f2f3f2;
    border: 1px solid #95999c;
    font-size: 22px;
    position: sticky;
    top: 60px;
    z-index: 2;

    &-wrapper {
      height: 175px;
      margin-top: -60px;
    }

    &::before {
      content: "";
      display: block;
      background: #fff;
      height: 60px;
      position: relative;
      top: -60px;
      border-left: 1px solid #333;
      margin: -1px;
    }
  }

  &-title-text {
    display: flex;
    align-items: center;
    justify-content: center;
    // margin-top: -60px;
    // flex: 1;
    height: 100%;

    &:not(&-static) {
      flex: 1;
    }

    &-static {
      border-right: 1px solid #95999c;
      width: 120px;
    }
  }

  &-body {
    // border: 1px solid #fff;
  }

  .unit {
    position: relative;
    border: 1px solid #95999c;
    margin: -1px 0 0 -1px;
    flex-grow: 1;
    padding: 0 15px;
    min-width: 160px;

    &-location {
      position: relative;
      width: 121px;
      border-bottom: 1px solid #95999c;
      border-left: 1px solid #95999c;
    }

    &-flag {
      position: absolute;
      left: 5px;
      top: 5px;
      width: 20px;
      height: 20px;
    }

    &-actions {
      position: absolute;
      left: 0;
      right: 0;
      top: 0;
      bottom: 0;
      display: none;
      background-color: white;

      .unit {
        &-action {
          cursor: pointer;
          &:hover {
            background-color: #f5f5f5;
          }
        }
      }
    }

    &:hover {
      opacity: 0.8;
      background-color: #f3f3f3;

      .unit-actions {
        display: flex;
      }
    }

    &-last-generation {
      height: 60px;
      flex: 1;
    }

    &-last-of-group:hover {
      box-shadow: 0 0 2px 1px var(--primary);
      border-color: #ececec;
    }

    &-new {
      cursor: pointer;
    }
  }
}
