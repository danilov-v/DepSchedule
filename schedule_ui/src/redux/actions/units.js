// UNITS
export const FETCH_UNITS = "FETCH_UNITS";
export const CREATE_UNIT = "CREATE_UNIT";
export const UPDATE_UNIT = "UPDATE_UNIT";
export const REMOVE_UNIT = "REMOVE_UNIT";
export const UPDATE_UNITS = "UPDATE_UNITS";

export const fetchUnits = payload => ({
  type: FETCH_UNITS,
  payload,
});

export const createUnit = payload => ({
  type: CREATE_UNIT,
  payload,
});

export const updateUnit = payload => ({
  type: UPDATE_UNIT,
  payload,
});

export const removeUnit = payload => ({
  type: REMOVE_UNIT,
  payload,
});

export const updateUnits = payload => ({
  type: UPDATE_UNITS,
  payload,
});
