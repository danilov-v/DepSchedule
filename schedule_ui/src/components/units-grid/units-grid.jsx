import React, { useState } from "react";
import PropTypes from "prop-types";
import { isNull, get } from "lodash";
import classnames from "classnames";
import { useSelector, useDispatch } from "react-redux";
import { getUnitRemoveConfirmationOptions } from "constants/confirmations";
import { MANAGE_UNITS } from "constants/permissions";
import { getAuthData } from "redux/selectors/auth";
import { removeUnit } from "redux/actions/units";
import { checkPermission } from "utils/permissions";
import { useConfirmation } from "components/confirmation-service/confirmation-service";

import { UnitPopup } from "./unit-popup";
import { UnitLocation } from "./unit-location";
import { Unit } from "./unit";
import { NewUnit } from "./new-unit";

import "./units-grid.scss";

const renderUnit = (
  { childUnit, title, unitId, flag, location },
  isTopLevel = true,
  isManageAble,
  handlers
) => {
  return childUnit && childUnit.length ? (
    <div
      key={unitId}
      className={classnames("d-flex", "flex-row-reverse", {
        "unit-last-of-group": isTopLevel,
      })}
    >
      <Unit
        key={unitId}
        title={title}
        unitId={unitId}
        {...handlers}
        isManageAble={isManageAble}
      />
      <div>
        {childUnit.map(unit => renderUnit(unit, false, isManageAble, handlers))}
      </div>
    </div>
  ) : (
    <div
      className={classnames("d-flex", {
        "unit-last-of-group": isTopLevel,
      })}
      key={unitId}
    >
      <UnitLocation flag={flag} location={location} />
      <Unit
        title={title}
        unitId={unitId}
        flag={flag}
        lastGen
        lastOfGroup={isTopLevel}
        isManageAble={isManageAble}
        {...handlers}
      />
    </div>
  );
};

export function UnitsGrid({ units, unitsTree, onUnitsUpdate }) {
  const dispatch = useDispatch();
  const confirm = useConfirmation();
  const [unitData, setUnitData] = useState(null);
  const { role } = useSelector(getAuthData);

  const onCloseEditPopup = () => {
    setUnitData(null);
  };

  const onAddUnit = unitParentId => {
    if (unitParentId) {
      setUnitData({ parentId: unitParentId, title: "" });
    } else {
      setUnitData(undefined);
    }
  };
  const onEditUnit = editingUnit => {
    const unit = units.find(unit => unit.unitId === editingUnit.unitId);

    setUnitData(unit);
  };
  const onRemoveUnit = unitId => {
    const unit = units.find(unit => unit.unitId === unitId);
    confirm(getUnitRemoveConfirmationOptions(unit)).then(() =>
      dispatch(removeUnit({ unitId }))
    );
  };

  const isManageAble = checkPermission(role, MANAGE_UNITS);

  return (
    <div className="units-grid">
      <div className="text-center units-grid-title">
        <div className="d-flex units-grid-title-wrapper">
          <div className="units-grid-title-text units-grid-title-text-static">
            Положение ПУ
          </div>
          <div className="units-grid-title-text">Подсистемы управления</div>
        </div>
      </div>
      <div className="units-grid-body">
        {unitsTree.map(unit =>
          renderUnit(unit, true, isManageAble, {
            onAddUnit,
            onEditUnit,
            onRemoveUnit,
          })
        )}
        <NewUnit hidden={!isManageAble} onAddUnit={onAddUnit} />
      </div>
      {!isNull(unitData) && (
        <UnitPopup
          units={units}
          unit={unitData}
          onClose={onCloseEditPopup}
          onUnitsUpdate={onUnitsUpdate}
          isEdit={Boolean(get(unitData, "title"))}
        />
      )}
    </div>
  );
}

UnitsGrid.propTypes = {
  units: PropTypes.arrayOf(PropTypes.object).isRequired,
  unitsTree: PropTypes.arrayOf(PropTypes.object).isRequired,
};
