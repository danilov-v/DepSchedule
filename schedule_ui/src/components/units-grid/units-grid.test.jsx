import React from "react";
import { shallow } from "enzyme";
import { UNITS } from "stub-data/units";

import { UnitsGrid } from "./units-grid";

describe("units-grid.jsx", () => {
  const testData = UNITS[1];

  describe("<UnitsGrid />", () => {
    it("should render and match snapshot", () => {
      const component = shallow(<UnitsGrid units={[testData]} />);

      expect(component).toMatchSnapshot();
    });
  });
});
