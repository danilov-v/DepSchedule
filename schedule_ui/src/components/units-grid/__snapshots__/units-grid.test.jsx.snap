// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`units-grid.jsx <UnitsGrid /> should render and match snapshot 1`] = `
<div
  className="units-grid"
>
  <div
    className="text-center units-grid-title"
  >
    Unit Grid Title
  </div>
  <div
    className="d-flex flex-row-reverse"
    key="2"
  >
    <Unit
      key="2"
      title="Система управления 2"
    />
    <div>
      <div
        className="d-flex flex-row-reverse"
        key="3"
      >
        <Unit
          key="3"
          title="Орган управления 2.1"
        />
        <div>
          <div
            className="d-flex flex-row-reverse"
            key="4"
          >
            <Unit
              key="4"
              title="Пункт управления 2.1.1"
            />
            <div>
              <Unit
                key="5"
                lastGen={true}
                title="Подразделение 2.1.1.1"
              />
              <Unit
                key="51"
                lastGen={true}
                title="Подразделение 2.1.1.2"
              />
            </div>
          </div>
        </div>
      </div>
      <Unit
        key="33"
        lastGen={true}
        title="Орган управления 2.2"
      />
    </div>
  </div>
</div>
`;
